File: legacy/python/agent/recommendation_agent.py
"""LLM-facing recommendation explanation helpers for GoTrends v2.

This module does not call an LLM. It prepares and validates the payload that an
LLM may explain, and provides a deterministic fallback explanation.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Any


@dataclass(frozen=True)
class RecommendationInput:
    campaign_id: int | str
    campaign_name: str
    recommended_action: str
    change_percent: float | None
    expected_incremental_cost: float | None
    expected_incremental_revenue: float | None
    expected_marginal_roas: float | None
    projected_cos: float | None
    confidence_score: int | None
    risk_level: str | None
    business_constraints_status: str
    constraints_reason: str | None
    approval_status: str
    reason: str | None


@dataclass(frozen=True)
class RecommendationExplanation:
    headline: str
    explanation: str
    expected_impact: str
    risk_and_confidence: str
    constraints_checked: str
    approval_note: str
    do_not_execute_reason: str | None


ACTION_LABELS = {
    "increase_budget": "aumentar budget",
    "increase_troas_or_reduce_budget": "aumentar tROAS ou reduzir budget",
    "optimize_efficiency": "otimizar eficiencia",
    "improve_ads_or_terms": "melhorar anuncios ou termos",
    "review_landing_or_offer": "revisar landing page ou oferta",
    "monitor": "monitorar",
}


def _fmt_currency(value: float | None) -> str:
    if value is None:
        return "indisponivel"
    return f"R$ {value:,.2f}".replace(",", "X").replace(".", ",").replace("X", ".")


def _fmt_pct(value: float | None) -> str:
    if value is None:
        return "indisponivel"
    return f"{value * 100:.1f}%".replace(".", ",")


def _fmt_number(value: float | None) -> str:
    if value is None:
        return "indisponivel"
    return f"{value:.2f}".replace(".", ",")


def build_llm_payload(row: dict[str, Any]) -> RecommendationInput:
    """Build the strict LLM input from a recommendation row."""
    required = {
        "campaign_id",
        "campaign_name",
        "recommended_action",
        "business_constraints_status",
        "approval_status",
    }
    missing = sorted(required - set(row))
    if missing:
        raise ValueError(f"Missing required recommendation fields: {', '.join(missing)}")

    return RecommendationInput(
        campaign_id=row["campaign_id"],
        campaign_name=row["campaign_name"],
        recommended_action=row["recommended_action"],
        change_percent=row.get("change_percent"),
        expected_incremental_cost=row.get("expected_incremental_cost"),
        expected_incremental_revenue=row.get("expected_incremental_revenue"),
        expected_marginal_roas=row.get("expected_marginal_roas"),
        projected_cos=row.get("projected_cos"),
        confidence_score=row.get("confidence_score"),
        risk_level=row.get("risk_level"),
        business_constraints_status=row["business_constraints_status"],
        constraints_reason=row.get("constraints_reason"),
        approval_status=row["approval_status"],
        reason=row.get("reason"),
    )


def explain_recommendation(payload: RecommendationInput) -> RecommendationExplanation:
    """Generate a deterministic explanation for review or LLM fallback."""
    action_label = ACTION_LABELS.get(payload.recommended_action, payload.recommended_action)
    blocked = payload.business_constraints_status == "blocked"
    review = payload.business_constraints_status == "needs_human_review"
    change_text = _fmt_pct(payload.change_percent)

    if blocked:
        headline = f"Acao bloqueada para {payload.campaign_name}"
        explanation = (
            f"A acao calculada foi {action_label}, mas ela nao deve seguir para execucao "
            f"porque o guardrail retornou bloqueio: {payload.constraints_reason or 'motivo nao informado'}."
        )
    else:
        headline = f"Candidata: {action_label} em {payload.campaign_name}"
        explanation = (
            f"A recomendacao calculada para a campanha {payload.campaign_name} e {action_label}. "
            f"A mudanca sugerida e {change_text}. Motivo estruturado: {payload.reason or 'indisponivel'}."
        )

    expected_impact = (
        "Impacto esperado: custo incremental "
        f"{_fmt_currency(payload.expected_incremental_cost)}, receita incremental "
        f"{_fmt_currency(payload.expected_incremental_revenue)}, ROAS marginal "
        f"{_fmt_number(payload.expected_marginal_roas)} e COS projetado "
        f"{_fmt_pct(payload.projected_cos)}."
    )
    risk_and_confidence = (
        f"Risco: {payload.risk_level or 'indisponivel'}. "
        f"Confianca estatistica: {payload.confidence_score if payload.confidence_score is not None else 'indisponivel'}."
    )
    constraints_checked = (
        f"Status dos guardrails: {payload.business_constraints_status}. "
        f"Razao: {payload.constraints_reason or 'indisponivel'}."
    )
    approval_note = (
        "Aprovacao humana pendente antes de qualquer execucao."
        if payload.approval_status == "pending" or review
        else f"Status de aprovacao: {payload.approval_status}."
    )
    do_not_execute_reason = payload.constraints_reason if blocked else None

    return RecommendationExplanation(
        headline=headline,
        explanation=explanation,
        expected_impact=expected_impact,
        risk_and_confidence=risk_and_confidence,
        constraints_checked=constraints_checked,
        approval_note=approval_note,
        do_not_execute_reason=do_not_execute_reason,
    )


def explanation_as_dict(explanation: RecommendationExplanation) -> dict[str, Any]:
    """Serialize explanation for logs or agent outputs."""
    return asdict(explanation)

File: app/src/agent/recommendationAgent.ts
// src/agent/recommendationAgent.ts
//
// LLM-facing recommendation explanation helpers for GoTrends v2.
//
// This module does NOT call an LLM. It validates the payload that an LLM may
// later explain, and provides a deterministic fallback explanation used as a
// baseline when the LLM is off.
//
// Ported 1:1 from `legacy/python/agent/recommendation_agent.py` (Task 2.11).
// Divergence vs. Python:
//   - Output field names use camelCase here (e.g. `expectedImpact`,
//     `doNotExecuteReason`) because this layer is LLM-facing and will be
//     serialized to JSON for a chat UI. Python uses snake_case
//     (`expected_impact`, `do_not_execute_reason`). `explanationAsDict`
//     emits the camelCase form on purpose — it is the contract.
//   - `buildLlmPayload` throws a plain `Error` (no ValueError equivalent in
//     TS), with a message mentioning the missing field(s).

export interface RecommendationInput {
  campaign_id: number | string
  campaign_name: string
  recommended_action: string
  change_percent: number | null
  expected_incremental_cost: number | null
  expected_incremental_revenue: number | null
  expected_marginal_roas: number | null
  projected_cos: number | null
  confidence_score: number | null
  risk_level: string | null
  business_constraints_status: string
  constraints_reason: string | null
  approval_status: string
  reason: string | null
}

export interface RecommendationExplanation {
  headline: string
  explanation: string
  expectedImpact: string
  riskAndConfidence: string
  constraintsChecked: string
  approvalNote: string
  doNotExecuteReason: string | null
}

export const ACTION_LABELS: Record<string, string> = {
  increase_budget: 'aumentar budget',
  increase_troas_or_reduce_budget: 'aumentar tROAS ou reduzir budget',
  optimize_efficiency: 'otimizar eficiencia',
  improve_ads_or_terms: 'melhorar anuncios ou termos',
  review_landing_or_offer: 'revisar landing page ou oferta',
  monitor: 'monitorar',
}

/**
 * Format a numeric value as Brazilian Real currency: "R$ 1.234,56".
 * Period as thousands separator, comma as decimal separator.
 * Null → "indisponivel".
 */
export function _fmt_currency(value: number | null): string {
  if (value === null || value === undefined) {
    return 'indisponivel'
  }
  // Mirror Python: f"R$ {value:,.2f}".replace(",", "X").replace(".", ",").replace("X", ".")
  const sign = value < 0 ? '-' : ''
  const abs = Math.abs(value)
  const fixed = abs.toFixed(2) // "1234.56"
  const [intPart = '0', fracPart = '00'] = fixed.split('.')
  const intWithThousands = intPart.replace(/\B(?=(\d{3})+(?!\d))/g, '.')
  return `R$ ${sign}${intWithThousands},${fracPart}`
}

/**
 * Format a fraction as percent with one decimal: 0.123 → "12,3%".
 * Null → "indisponivel".
 */
export function _fmt_pct(value: number | null): string {
  if (value === null || value === undefined) {
    return 'indisponivel'
  }
  // Mirror Python: f"{value*100:.1f}%".replace(".", ",")
  return `${(value * 100).toFixed(1)}%`.replace('.', ',')
}

/**
 * Format a number with two decimals: 2.98 → "2,98".
 * Null → "indisponivel".
 */
export function _fmt_number(value: number | null): string {
  if (value === null || value === undefined) {
    return 'indisponivel'
  }
  // Mirror Python: f"{value:.2f}".replace(".", ",")
  return value.toFixed(2).replace('.', ',')
}

/**
 * Validate a raw recommendation row and produce a strict RecommendationInput.
 * Throws if required fields are missing.
 */
export function buildLlmPayload(row: Record<string, unknown>): RecommendationInput {
  const required = [
    'campaign_id',
    'campaign_name',
    'recommended_action',
    'business_constraints_status',
    'approval_status',
  ] as const

  const missing = required.filter((k) => !(k in row)).sort()
  if (missing.length > 0) {
    throw new Error(`Missing required recommendation fields: ${missing.join(', ')}`)
  }

  const getNum = (k: string): number | null => {
    const v = row[k]
    if (v === undefined || v === null) return null
    if (typeof v === 'number') return v
    throw new Error(`Field ${k} must be a number or null, got ${typeof v}`)
  }
  const getStr = (k: string): string | null => {
    const v = row[k]
    if (v === undefined || v === null) return null
    if (typeof v === 'string') return v
    throw new Error(`Field ${k} must be a string or null, got ${typeof v}`)
  }

  const campaignId = row.campaign_id
  if (typeof campaignId !== 'number' && typeof campaignId !== 'string') {
    throw new Error('Field campaign_id must be number or string')
  }
  const campaignName = row.campaign_name
  if (typeof campaignName !== 'string') {
    throw new Error('Field campaign_name must be a string')
  }
  const recommendedAction = row.recommended_action
  if (typeof recommendedAction !== 'string') {
    throw new Error('Field recommended_action must be a string')
  }
  const businessConstraintsStatus = row.business_constraints_status
  if (typeof businessConstraintsStatus !== 'string') {
    throw new Error('Field business_constraints_status must be a string')
  }
  const approvalStatus = row.approval_status
  if (typeof approvalStatus !== 'string') {
    throw new Error('Field approval_status must be a string')
  }

  return {
    campaign_id: campaignId,
    campaign_name: campaignName,
    recommended_action: recommendedAction,
    change_percent: getNum('change_percent'),
    expected_incremental_cost: getNum('expected_incremental_cost'),
    expected_incremental_revenue: getNum('expected_incremental_revenue'),
    expected_marginal_roas: getNum('expected_marginal_roas'),
    projected_cos: getNum('projected_cos'),
    confidence_score: getNum('confidence_score'),
    risk_level: getStr('risk_level'),
    business_constraints_status: businessConstraintsStatus,
    constraints_reason: getStr('constraints_reason'),
    approval_status: approvalStatus,
    reason: getStr('reason'),
  }
}

/**
 * Deterministic explanation used as the LLM-off baseline.
 */
export function explainRecommendation(payload: RecommendationInput): RecommendationExplanation {
  const actionLabel = ACTION_LABELS[payload.recommended_action] ?? payload.recommended_action
  const blocked = payload.business_constraints_status === 'blocked'
  const review = payload.business_constraints_status === 'needs_human_review'
  const changeText = _fmt_pct(payload.change_percent)

  let headline: string
  let explanation: string
  if (blocked) {
    headline = `Acao bloqueada para ${payload.campaign_name}`
    explanation =
      `A acao calculada foi ${actionLabel}, mas ela nao deve seguir para execucao ` +
      `porque o guardrail retornou bloqueio: ${payload.constraints_reason ?? 'motivo nao informado'}.`
  } else {
    headline = `Candidata: ${actionLabel} em ${payload.campaign_name}`
    explanation =
      `A recomendacao calculada para a campanha ${payload.campaign_name} e ${actionLabel}. ` +
      `A mudanca sugerida e ${changeText}. Motivo estruturado: ${payload.reason ?? 'indisponivel'}.`
  }

  const expectedImpact =
    'Impacto esperado: custo incremental ' +
    `${_fmt_currency(payload.expected_incremental_cost)}, receita incremental ` +
    `${_fmt_currency(payload.expected_incremental_revenue)}, ROAS marginal ` +
    `${_fmt_number(payload.expected_marginal_roas)} e COS projetado ` +
    `${_fmt_pct(payload.projected_cos)}.`

  const riskAndConfidence =
    `Risco: ${payload.risk_level ?? 'indisponivel'}. ` +
    `Confianca estatistica: ${payload.confidence_score !== null && payload.confidence_score !== undefined ? payload.confidence_score : 'indisponivel'}.`

  const constraintsChecked =
    `Status dos guardrails: ${payload.business_constraints_status}. ` +
    `Razao: ${payload.constraints_reason ?? 'indisponivel'}.`

  const approvalNote =
    payload.approval_status === 'pending' || review
      ? 'Aprovacao humana pendente antes de qualquer execucao.'
      : `Status de aprovacao: ${payload.approval_status}.`

  const doNotExecuteReason = blocked ? payload.constraints_reason : null

  return {
    headline,
    explanation,
    expectedImpact,
    riskAndConfidence,
    constraintsChecked,
    approvalNote,
    doNotExecuteReason,
  }
}

/**
 * Serialize an explanation for logs or agent outputs.
 *
 * Field names are camelCase, matching the TS `RecommendationExplanation`
 * interface. Python's `explanation_as_dict` emits snake_case; this is the
 * deliberate divergence documented at the top of the file.
 */
export function explanationAsDict(
  explanation: RecommendationExplanation,
): Record<string, unknown> {
  return {
    headline: explanation.headline,
    explanation: explanation.explanation,
    expectedImpact: explanation.expectedImpact,
    riskAndConfidence: explanation.riskAndConfidence,
    constraintsChecked: explanation.constraintsChecked,
    approvalNote: explanation.approvalNote,
    doNotExecuteReason: explanation.doNotExecuteReason,
  }
}
